src/extractor/node_based_graph_factory.cpp

```diff
@@ -174,11 +183,23 @@ void NodeBasedGraphFactory::CompressAnnotationData()
         for (EdgeID nbg_edge_id : compressed_output_graph.GetAdjacentEdgeRange(nbg_node_u))
         {
             auto &edge = compressed_output_graph.GetEdgeData(nbg_edge_id);
-            edge.annotation_data = annotation_mapping[edge.annotation_data];
-            BOOST_ASSERT(edge.annotation_data != INVALID_ANNOTATIONID);
+            auto const it = unique_annotations.find(edge.annotation_data);
+            BOOST_ASSERT(it != unique_annotations.end());
+            auto const it2 = cached_ids.find(*it);
+            BOOST_ASSERT(it2 != cached_ids.end());
+
+            edge.annotation_data = it2->second;
         }
     }
 
+    // mark unused references for remove
+    for (AnnotationID id = 0; id < annotation_data.size(); ++id)
+    {
+        auto const it = unique_annotations.find(id);
+        if (it == unique_annotations.end() || *it != id)
```

Comment:
just curious how `*it != id` is possible here for `AnnotationID`?

Reply:
This happens, since we don't compare based on ID but based on annotation_data content. So there are multiple IDS (like 1,2,3) which are equal, but we only have a single representation for them (e.g. 3)

Reply:
Set compares by real data. And store any unique id for each equal data. In this line we mark for remove all other duplicates and save just this one (unique)

Reply:
ok, but it is just another source of non-determinism: `{1,2,3}` also a posible set

Reply:
yes, and thats not a problem for algorithm, because in that case we also remove just duplicated annotation_data values (with the same data values)

Reply:
as per chat: it is not a problem in the implementation

Reply:
To extend on the reasoning here: the implementation always choses the first ID within the graph as a representative. There is no randomness involved in the choosing process, making this deterministic.